docs/design.md

```diff
@@ -0,0 +1,118 @@
+# Design Notes
+
+The current goals of the Hypothesis for Ruby project are:
+
+* To provide a useful but not wholly feature complete version of
+  [Hypothesis](http://hypothesis.works) for Ruby, that works with
+  RSpec (and ideally minitest, but if that at any point proves to
+  be a lot of work this may be dropped. It's not an explicit
+  requirement, but supporting it now makes it much easier to find
+  the right shape of the project design).
+* To provide a mostly feature complete version of the Conjecture
+  engine that powers Hypothesis in Rust, as decoupled from that
+  Ruby front-end as possible.
+
+Hypothesis for Ruby is not intended to be an exact feature for
+feature copy of the Python version. It will have a lot of the same
+underlying functionality, but with a number of changes driven by:
+
+* Trying to make it feel as "ruby native" as possible.
+* The ability to design an API from scratch that lacks many of the
+  constraints imposed both by the earlier much more limited functionality
+  of Hypothesis and the specifics of Python decorators and test
+  frameworks.
+
+## Differences
+
+The most fundamental API differences  between Hypothesis
+for Python and Hypothesis for Ruby are:
+
+* In Python we do a whole giant song and dance about exposing
+  functions for the test runner to call, while in Ruby we just
+  have a function which repeatedly calls a block and then fails.
+* in Python you specify a bunch of given parameters up front,
+  and then if you want values inline in the test you [explicitly
+  opt in to it](https://hypothesis.readthedocs.io/en/latest/data.html#drawing-interactively-in-tests),
+  while in Ruby this is not only the default but the only way to
+  get those values.
+* Strategies are called providers because strategy is a terrible
+  name that was originally intended to be internal and then leaked
+  into the public API because I wasn't thinking hard about naming.
+
+So for example:
+
+```ruby
+RSPec.describe "integer addition" do
+  it "commutes" do
+    hypothesis do
+      m = given integers
```

Comment:
The singular/plural mismatch here throws me a bit. I'd kinda expect either:

```ruby
ms = given integers
```

or something that makes more sense with a singular `m`, like:

```ruby
m = from integers
```

or

```ruby
m = given integer
```

Reply:
If you go this way I think you should make the plural alias not have any articles, so `integer` and `integers`. Then you could make `a` and `an` noop messages to `given`, leaving you with `given an integer`.

Reply:
Do you mean the singular alias?

As per comment below I've gone with `any_integer`. `an/a` doesn't really read right but...

Hmm...

How about `any` as a replacement name for `given`? It reads a lot better I think...